protocol.go

```diff
@@ -48,6 +48,45 @@ const (
 
 var errNoTimeout = errors.New("no timeout")
 
+// InferProtocolFromRequest returns the inferred protocol name for parsing an
+// HTTP request. It inspects the request's method and headers to determine the
+// protocol. If the request doesn't match any known protocol, an empty string
+// is returned.
+func InferProtocolFromRequest(request *http.Request) string {
```

Comment:
- The rest of the API uses For and not From
- would this be better named "ProtocolForRequest"?
- would it be better to return string, bool?

Reply:
The `For` is used to distinguish streams that are to be consumed by the client rather than the handler. So I think the `From`  is better here even tho its not currently in the API.

Reply:
In my mind they are not necessarily synonyms where you'd always use one over another for consistency.
In my mind, the distinction is that `For` in a function name is used when the function constructs a new value that corresponds to its input or even wraps its input; `From` is used when the function extracts a value out of the inputs. In that case, `From` seems appropriate to me.

As far as returning `string` vs. `(string, bool)`: another option is to add a `ProtocolUnknown` constant, so that callers can use a switch and don't need to examine two values. If that approach sounds good, the constant value might need to be the empty string so that an uninitialized `Peer` object reports `ProtocolUnknown`. WDYT?